cub/cub/device/device_reduce.cuh

```diff
@@ -557,38 +565,53 @@ struct DeviceReduce
     size_t& temp_storage_bytes,
     InputIteratorT d_in,
     OutputIteratorT d_out,
-    int num_items,
+    ::cuda::std::int64_t num_items,
     cudaStream_t stream = 0)
   {
     CUB_DETAIL_NVTX_RANGE_SCOPE_IF(d_temp_storage, "cub::DeviceReduce::ArgMin");
 
-    // Signed integer type for global offsets
-    using OffsetT = int;
-
     // The input type
     using InputValueT = cub::detail::value_t<InputIteratorT>;
 
-    // The output tuple type
-    using OutputTupleT = cub::detail::non_void_value_t<OutputIteratorT, KeyValuePair<OffsetT, InputValueT>>;
+    // Offset type used within the kernel and to index within one partition
+    using PerPartitionOffsetT = int;
 
-    using AccumT = OutputTupleT;
+    // Offset type used to index within the total input in the range [d_in, d_in + num_items)
+    using GlobalOffsetT = ::cuda::std::int64_t;
 
-    using InitT = detail::reduce::empty_problem_init_t<AccumT>;
+    // Determine initial value type
+    using OutputTupleT = cub::detail::non_void_value_t<OutputIteratorT, KeyValuePair<GlobalOffsetT, InputValueT>>;
+    using InitT        = typename OutputTupleT::Value;
 
-    // The output value type
-    using OutputValueT = typename OutputTupleT::Value;
-
-    // Wrapped input iterator to produce index-value <OffsetT, InputT> tuples
-    using ArgIndexInputIteratorT = ArgIndexInputIterator<InputIteratorT, OffsetT, OutputValueT>;
-
-    ArgIndexInputIteratorT d_indexed_in(d_in);
+    // Reduction operation
+    using ReduceOpT = cub::ArgMin;
 
     // Initial value
     // TODO Address https://github.com/NVIDIA/cub/issues/651
-    InitT initial_value{AccumT(1, Traits<InputValueT>::Max())};
-
-    return DispatchReduce<ArgIndexInputIteratorT, OutputIteratorT, OffsetT, cub::ArgMin, InitT, AccumT>::Dispatch(
-      d_temp_storage, temp_storage_bytes, d_indexed_in, d_out, num_items, cub::ArgMin(), initial_value, stream);
+    InitT initial_value{Traits<InputValueT>::Max()};
+
+    // Helper transform output iterator, to allow "implicit conversion" between KeyValuePair types that have a different
+    // key type, which may happen if the user uses a different index type than the global offset type used by the
+    // algorithm
+    using implicit_cast_kv_pair_op_it =
+      THRUST_NS_QUALIFIER::tabulate_output_iterator<detail::reduce::write_to_user_out_it<OutputIteratorT>>;
+    implicit_cast_kv_pair_op_it out_it =
+      THRUST_NS_QUALIFIER::make_tabulate_output_iterator(detail::reduce::write_to_user_out_it<OutputIteratorT>{d_out});
+
+    return detail::reduce::dispatch_streaming_arg_reduce_t<
+      InputIteratorT,
+      implicit_cast_kv_pair_op_it,
```

Comment:
Suggestion: maybe simplifies the code.
```suggestion
    auto out_it =
      THRUST_NS_QUALIFIER::make_tabulate_output_iterator(detail::reduce::write_to_user_out_it<OutputIteratorT>{d_out});

    return detail::reduce::dispatch_streaming_arg_reduce_t<
      InputIteratorT,
      decltype(out_it),
```
Applies a second time below.

Reply:
Nice simplification. Thanks.